src/MCPClient/lib/archivematicaClient.py

```diff
@@ -21,166 +47,190 @@
+    script = client_script_full_path + ' ' + arguments
+    return script, task_uuid, capture_output
 
 
 @auto_close_db
-def executeCommand(gearman_worker, gearman_job):
+def execute_command(gearman_worker, gearman_job):
+    """Execute the command encoded in ``gearman_job`` and return its exit code,
+    standard output and standard error as a pickled dict.
+    """
     try:
-        execute = gearman_job.task
-        logger.info('Executing %s (%s)', execute, gearman_job.unique)
-        data = cPickle.loads(gearman_job.data)
-        utcDate = getUTCDate()
-        arguments = data["arguments"]  # .encode("utf-8")
-        if isinstance(arguments, unicode):
-            arguments = arguments.encode("utf-8")
-
-        sInput = ""
-        clientID = gearman_worker.worker_client_id
-
-        task = Task.objects.get(taskuuid=gearman_job.unique)
-        if task.starttime is not None:
-            exitCode = -1
-            stdOut = ""
-            stdError = """Detected this task has already started!
-Unable to determine if it completed successfully."""
-            return cPickle.dumps({"exitCode": exitCode, "stdOut": stdOut, "stdError": stdError})
-        else:
-            task.client = clientID
-            task.starttime = utcDate
-            task.save()
-
-        if execute not in supportedModules:
-            output = ["Error!", "Error! - Tried to run and unsupported command."]
-            exitCode = -1
-            return cPickle.dumps({"exitCode": exitCode, "stdOut": output[0], "stdError": output[1]})
-        command = supportedModules[execute]
-
-        replacementDic["%date%"] = utcDate.isoformat()
-        replacementDic["%jobCreatedDate%"] = data["createdDate"]
-        # Replace replacement strings
-        for key in replacementDic.keys():
-            command = command.replace(key, replacementDic[key])
-            arguments = arguments.replace(key, replacementDic[key])
-
-        key = "%taskUUID%"
-        value = gearman_job.unique.__str__()
-        arguments = arguments.replace(key, value)
-
-        # Execute command
-        command += " " + arguments
-        logger.info('<processingCommand>{%s}%s</processingCommand>', gearman_job.unique, command)
-        exitCode, stdOut, stdError = executeOrRun(
-            "command", command, sInput, printing=True, capture_output=False)
-        return cPickle.dumps({"exitCode": exitCode, "stdOut": stdOut, "stdError": stdError})
+        script, task_uuid, capture_output = _process_gearman_job(
+            gearman_job, gearman_worker)
+    except ProcessGearmanJobError as error:
+        return cPickle.dumps(error)
     except OSError:
         logger.exception('Execution failed')
-        output = ["Archivematica Client Error!", traceback.format_exc()]
-        exitCode = 1
-        return cPickle.dumps({"exitCode": exitCode, "stdOut": output[0], "stdError": output[1]})
+        return cPickle.dumps({'exitCode': 1,
```

Comment:
^ this `OSError` was used to control the exceptions raised by subprocess (executeOrRun), should be done inside _process_gearman_job instead?